### ISSUE TITLE
callback.AverageMetric.on_batch_end() last_output not sent as list

### ISSUE THREAD

**Issue by plynch-chewy** [fastai/fastai#1145]
https://github.com/fastai/fastai/blob/2f9697ac4dec048a43dbb0a16ec85334ce95b069/fastai/callback.py#L268:L271

On the function linked above, the `last_output` isn't checked for/made into a list. This becomes a problem if the model's forward pass returns a list. Below is an example of case:

```
from fastai import *
from fastai.train import *
import torch as t

# sequence of ids 1 to 9 with context (0 as padding)
data = [
    [1, [0, 0, 2, 3]],
    [2, [0, 1, 3, 4]],
    [3, [1, 2, 4, 5]],
    [4, [2, 3, 5, 6]],
    [5, [3, 4, 6, 7]],
    [6, [4, 5, 7, 8]],
    [7, [5, 6, 8, 9]],
    [8, [6, 7, 9, 0]],
    [9, [7, 8, 0, 0]]]

class SeqDataset(Dataset):
    def __init__(self, data): self.data = data
    def __len__(self): return len(self.data)
    def __getitem__(self, idx):
        inp, ctx = self.data[idx]
        return inp, np.array(ctx)

dl_trn = DataLoader(SeqDataset(data), batch_size=2)
dl_val = DataLoader(SeqDataset(data), batch_size=2*2)

def collate_wrapper(samples):
    # need to wrap inps in [] since xb gets input into loss_fn as *xb
    inps = [t.tensor([s[i] for s in samples]) for i in range(len(samples[0]))]
    # add a dummy yb
    return [[inps], torch.zeros(len(samples))]

databunch = DataBunch(dl_trn, dl_val, device='cpu', collate_fn=collate_wrapper)

# skip gram with negative sampling https://arxiv.org/pdf/1310.4546.pdf
class SGNS(nn.Module):
    def __init__(self, inp_size, emb_dim=3, n_negs=2):
        super().__init__()
        self.inp_size = inp_size
        self.emb_dim = emb_dim
        self.embs = nn.Embedding(inp_size, emb_dim)
        #self.embs = layers.get_embedding(inp_size, emb_dim)
        self.n_negs = n_negs

    def sample_negs(self, bs):
        return t.FloatTensor(bs, self.n_negs).uniform_(0, self.inp_size - 1).long()

    def forward(self, inps):
        inp, ctx = inps
        bs = inp.size()[0]
        ctx_size = ctx.size()[1]
        negs = self.sample_negs(bs)
        out = [
            self.embs(inp).unsqueeze(2),
            self.embs(ctx),
            self.embs(negs).neg()]
        return out
    
    def cosine_sim_loss(self, vecs_a, vecs_b):
        return -(t.bmm(vecs_b, vecs_a).squeeze(2).sigmoid().log().mean(1)).mean()

    def loss_ctx(self, inp_vecs, ctx_vecs): return self.cosine_sim_loss(inp_vecs, ctx_vecs)
    def loss_neg(self, inp_vecs, neg_vecs): return self.cosine_sim_loss(inp_vecs, neg_vecs)

    def loss(self, inps, _):
        inp_vecs, ctx_vecs, neg_vecs = inps
        loss_all = self.loss_ctx(inp_vecs, ctx_vecs) + self.loss_neg(inp_vecs, neg_vecs)
        return loss_all

m = SGNS(inp_size=10, emb_dim=3, n_negs=2)

# works
learn = Learner(databunch, m, loss_func=m.loss)
learn.fit(1)

# breaks
metrics = [m.loss_ctx, m.loss_neg]
learn_with_metrics = Learner(databunch, m, loss_func=m.loss, metrics=metrics)
learn_with_metrics.fit(1)
```

Since my loss is a combination of context words and negative words, I wanted to print out metrics on both at end of epoch. The main problem is that I'm returning a list with my forward pass. Perhaps I could code this differently? But since inputs are passed to the model as `m(*xb)` in the fit function but that output is then passed to without star notation like `self.func(last_output, *last_target)`, I can't figure out a way. I expect making `last_output` into a list and using star notation would make this functionality work like other parts in the library, but I'm not confident that it won't break other pieces of code.

Here would be my change:
```
def on_batch_end(self, last_output, last_target, train, **kwargs):
    if not is_listy(last_output): last_output=[last_output]
    if not is_listy(last_target): last_target=[last_target]
    self.count += last_target[0].size(0)
    self.val += last_target[0].size(0) * self.func(*last_output, *last_target).detach().cpu()
```

**Comment by sgugger** [comment]
Not sure why this is an issue: your model can return a list of outputs but it's always treated as one argument in the rest of the code. Why not adapt your function to extract the things you want from this list of outputs?

**Comment by plynch-chewy** [comment]
Ahh ok, thanks. I made the mistake of the different losses not accepting the same type of parameters.

This worked for me:
```
    def loss_ctx(self, inps, _):
        inp_vecs, ctx_vecs, neg_vecs = inps
        return self.cosine_sim_loss(inp_vecs, ctx_vecs)
    def loss_neg(self, inps, _):
        inp_vecs, ctx_vecs, neg_vecs = inps
        return self.cosine_sim_loss(inp_vecs, neg_vecs)
    def loss(self, inps, _):
        loss_all = self.loss_ctx(inps,_) + self.loss_neg(inps,_)
        return loss_all
```

Thanks!
